lib/waive.py

```diff
@@ -0,0 +1,215 @@
+"""
+Functionality for an automatic failure waiving logic, configured by
+a custom file format. See WAIVES.md.
+"""
+
+import os
+import re
+import textwrap
+from pathlib import Path
+
+import util
+import versions
+
+_sections_cache = None
+
+WAIVE_FILE = Path(util.libdir).parent / 'conf' / 'waives'
+
+
+class _PushbackIterator:
+    """
+    A simple wrapper for an iterator that allows the user to "push back"
+    the last retrieved item, so it appears again in next().
+    """
+    def __init__(self, source):
+        self.it = iter(source)
+        self.pushed_back = False
+        self.last = None
+        self.counter = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.pushed_back:
+            self.pushed_back = False
+            return self.last
+        else:
+            new = next(self.it)
+            self.last = new
+            self.counter += 1
+            return new
+
+    def pushback(self):
+        self.pushed_back = True
+
+
+class WaiveParseError(SyntaxError):
+    """
+    Easy waive file syntax error reporting, with line numbers derived
+    from _PushbackIterator style counter.
+    """
+    def __init__(self, meta, msg):
+        super().__init__(f"waive line {meta.counter}: {msg}")
+
+
+def _compile_eval(meta, code):
+    if not code.strip():
+        raise WaiveParseError(meta, "empty python block code ending here")
+    try:
+        return compile(textwrap.dedent(code), 'waivecode', 'eval')
+    except Exception:
+        raise WaiveParseError(meta, "compiling waive python code failed")
+
+
+def _parse_waive_file(stream):
+    sections = []
+    regexes = set()
+    python_code = ''
+    state = 'skipping_empty_lines'
+
+    lines = _PushbackIterator(stream)
+    for line in lines:
+        if line.startswith('#'):
+            continue
+        line = line.rstrip('\n')
+
+        if state == 'skipping_empty_lines':
+            if line:
+                state = 'reading_regex'
+                lines.pushback()
+                continue
+
+        elif state == 'reading_regex':
+            if not line:
+                raise WaiveParseError(lines, "unexpected empty line between regexes")
+
+            if line.startswith((' ', '\t')):
+                if not regexes:
+                    raise WaiveParseError(lines, "python block without a preceding regexp")
+                state = 'reading_python'
+                lines.pushback()
+                continue
+
+            try:
+                regexes.add(re.compile(line))
+            except re.error as e:
+                raise WaiveParseError(lines, f"regex failed: {e}")
+
```

Comment:
Just a suggestion to make the logic there more obvious. It took me some time to get the point in the original code so I tried to come up with something simpler, but it still doesn't feel right :/
```suggestion
            if not line.startswith((' ', '\t')):
                try:
                    regexes.add(re.compile(line))
                except re.error as e:
                    raise WaiveParseError(lines, f"regex failed: {e}")
                continue
            else:
                if regexes:
                    state = 'reading_python'
                    lines.pushback()
                else:
                    raise WaiveParseError(lines, "python block without a preceding regexp")
```